packages/eslint-plugin/tests/rules/no-floating-promises.test.ts

```diff
@@ -1117,5 +1168,264 @@ async function test() {
         },
       ],
     },
+    {
+      code: `
+async function foo() {
+  const myPromise = async () => void 0;
+  const condition = true;
+
+  void condition || myPromise();
+}
+      `,
+      errors: [
+        {
+          line: 6,
+          messageId: 'floatingVoid',
+          suggestions: [
+            {
+              messageId: 'floatingFixVoid',
+              output: `
+async function foo() {
+  const myPromise = async () => void 0;
+  const condition = true;
+
+  void (void condition || myPromise());
+}
+      `,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
+async function foo() {
+  const myPromise = async () => void 0;
+  const condition = true;
+
+  (await condition) && myPromise();
+}
+      `,
+      options: [{ ignoreVoid: false }],
+      errors: [
+        {
+          line: 6,
+          messageId: 'floating',
+          suggestions: [
+            {
+              messageId: 'floatingFixAwait',
+              output: `
+async function foo() {
+  const myPromise = async () => void 0;
+  const condition = true;
+
+  await ((await condition) && myPromise());
+}
+      `,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
+async function foo() {
+  const myPromise = async () => void 0;
+  const condition = true;
+
+  condition && myPromise();
+}
+      `,
+      errors: [
+        {
+          line: 6,
+          messageId: 'floatingVoid',
+          suggestions: [
+            {
+              messageId: 'floatingFixVoid',
+              output: `
+async function foo() {
+  const myPromise = async () => void 0;
+  const condition = true;
+
+  void (condition && myPromise());
```

Comment:
The entire logical expression is enclosed in parentheses as in return-await.
https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=IYZwngdgxgBAZgV2gFwJYHsIwEbAE4AUAlDAN4wC+AsAFCiSyIoZZzrrFm0wzJ5hcaPHngCmyBHixIAJqLioIomTAA+qnPmIBubpRhRgyKAAsylWhSA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRfWaJ2gQwDufSvnRRE0aAHtokcGAC+IBUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA